api/v1alpha1/openstackimage_types.go

```diff
@@ -0,0 +1,218 @@
+type ImageHashAlgorithm string
+
+const (
+	ImageHashAlgorithmMD5    ImageHashAlgorithm = "md5"
+	ImageHashAlgorithmSHA1   ImageHashAlgorithm = "sha1"
+	ImageHashAlgorithmSHA256 ImageHashAlgorithm = "sha256"
+	ImageHashAlgorithmSHA512 ImageHashAlgorithm = "sha512"
+)
+
+// +kubebuilder:validation:XValidation:rule="!in(self, ['created_at', 'updated_at', 'status', 'checksum', 'size', 'virtual_size', 'direct_url', 'self', 'file', 'schema', 'id', 'os_hash_algo', 'os_hash_value', 'location', 'deleted', 'deleted_at', 'container_format', 'disk_format', 'min_disk', 'min_ram', 'name', 'tags', 'owner', 'visibility', 'protected', 'os_hidden'])"
+type ImageAdditionalPropertyName string
+
+type ImageAdditionalProperty struct {
+	// Name is the name of the glance property. It is an error for it to conflict with any explicitly defined property.
+	Name ImageAdditionalPropertyName `json:"name"`
+
+	// Value is the value of the glance property
+	// +required
+	// +kubebuilder:validation:MinLength:=1
+	// +kubebuilder:validation:MaxLength:=1024
+	Value string `json:"value"`
+}
+
+type ImageProperties struct {
+	// MinDisk is the minimum amount of disk space in GB that is required to boot the image
+	// +kubebuilder:validation:Minimum:=1
+	MinDiskGB *int `json:"minDiskGB,omitempty"`
+
+	// MinRAMMB is the minimum amount of RAM in MB that is required to boot the image.
+	// +kubebuilder:validation:Minimum:=1
+	MinRAMMB *int `json:"minRAMMB,omitempty"`
+}
+
+// +kubebuilder:validation:Enum:=xz;gz
+type ImageCompression string
+
+const (
+	ImageCompressionXZ ImageCompression = "xz"
+	ImageCompressionGZ ImageCompression = "gz"
+)
+
+type ImageContent struct {
+	// Hash is a hash which can be used to verify the downloaded image data
+	Hash *ImageHash `json:"hash,omitempty"`
+
+	// GlanceHashAlgorithm is the algorithm of the hash glance will publish.
+	// If not set it will default to sha512, which is Glance's default. It
+	// MUST be set to the value configured in Glance.
+	// +kubebuilder:default:=sha512
+	// +required
+	GlanceHashAlgorithm *ImageHashAlgorithm `json:"glanceHashAlgorithm,omitempty"`
+
+	// Source specifies how to obtain the image data
+	Source ImageSource `json:"source"`
+}
+
+type ImageSourceType string
+
+const (
+	ImageSourceTypeURL ImageSourceType = "url"
+)
+
+// ImageSource specifies the source of image data
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'url' ?  has(self.url) : !has(self.url)",message="url is required when type is url, and forbidden otherwise"
+// +union
+type ImageSource struct {
+	// Type is the type of the image source
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	Type ImageSourceType `json:"type"`
+
+	// URL describes how to obtain image data by downloading it from a URL. Must be set if Type is 'url'
+	// +unionMember
+	URL *ImageSourceURL `json:"url,omitempty"`
+}
+
+type ImageSourceURL struct {
+	// URL containing image data
+	// +kubebuilder:validation:Format=uri
```

Comment:
Use CEL HTTP library to validate that this is a valid URL? Don't think this actually works

Reply:
I wasn't sure either, but it kinda looked like it would and I'm going to write tests for it anyway.

I couldn't work out how to write this in CEL. Got an example?

Reply:
You probably want something like below, that's very basic, but, if you prefer, you can then get creative and expand.
```
// +kubebuilder:validation:XValidation:rule="isURL(self)",message="url must be a valid URL"
```

If you need more validation the following are also available:
> Once parsed via the `url` function, the resulting URL object has `getScheme`, `getHost`, `getHostname`, `getPort`, `getEscapedPath` and `getQuery` accessors.